src/filter_tree/filter_tree.c

```diff
@@ -612,9 +628,76 @@ static bool _FilterTree_Compact_Or(FT_FilterNode *node) {
 	}
 }
 
+// Compacts 'XOR' and 'XNOR' condition nodes.
+static bool _FilterTree_Compact_XOr(FT_FilterNode *node, bool xnor) {
+	// try to compact left and right children
+	bool is_lhs_const = FilterTree_Compact(node->cond.left);
+	bool is_rhs_const = FilterTree_Compact(node->cond.right);
+	// if both are not compactable, this node is not compactable
+	if(!is_lhs_const && !is_rhs_const) return false;
+
+	// in every case from now, there will be a reduction,
+	// save the children in local placeholders for current node in-place modifications
+	bool final_value = false;
+	FT_FilterNode *lhs = node->cond.left;
+	FT_FilterNode *rhs = node->cond.right;
+	// both children are constants. This node can be set as constant expression
+	if(is_lhs_const && is_rhs_const) {
+		// both children are now contant expressions, evaluate and compact
+		final_value = SIValue_IsTrue(AR_EXP_Evaluate(rhs->exp.exp, NULL));
+		if(final_value) {
+			final_value = SIValue_IsFalse(AR_EXP_Evaluate(lhs->exp.exp, NULL));
+		} else {
+			final_value = SIValue_IsTrue(AR_EXP_Evaluate(lhs->exp.exp, NULL));
+		}
+
+		// invert the result if we are performing XNOR
+		if(xnor) final_value = !final_value;
+
+		// final value is XOR operation on lhs and rhs - reducing an OR node
+		// in place set the node to be an expression node
+		_FilterTree_In_Place_Set_Exp(node, SI_BoolVal(final_value));
+		FilterTree_Free(lhs);
+		FilterTree_Free(rhs);
+		return true;
+	} else {
+		// only one of the nodes is constant, find and evaluate
```

Comment:
I don't think we can reduce the root node (`xor`, `xnor`) in this `else` case, the best we can do is reduce the reducible LHS or RHS sub tree.

Reply:
done